runtime/vm/classallocation.c

```diff
@@ -377,7 +377,7 @@ freeClassLoader(J9ClassLoader *classLoader, J9JavaVM *javaVM, J9VMThread *vmThre
 	}
 
 	if (NULL != classLoader->jniIDs) {
-		if (J9_ARE_ANY_BITS_SET(javaVM->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_KEEP_JNI_IDS)) {
+		if (javaVM->asyncGetCallTraceUsed || J9_ARE_ANY_BITS_SET(javaVM->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_KEEP_JNI_IDS)) {
```

Comment:
Wondering if we should have a way to disable this behaviour. Perhaps we should turn on `J9_EXTENDED_RUNTIME2_KEEP_JNI_IDS` by default and make this `javaVM->asyncGetCallTraceUsed && J9_ARE_ANY_BITS_SET(javaVM->extendedRuntimeFlags2, J9_EXTENDED_RUNTIME2_KEEP_JNI_IDS)`.

I dont see a reason to keep JNIIds if one is not using AGCT

Reply:
If we automatically do the retention for ASGCT, then the disabling option would be to favour the occasional crash over the memory leak. We've seen this request before in the past, so it's not unreasonable to do this if the crashing is infrequent.

The only other reason I can think of to maintain the existing command line option is as a debug feature for service (to prove that IDs are being used incorrectly).

At the cost of another runtime flag bit, we could make:

No option specified - keep IDs only if ASGCT is used
-XX:+KeepJNIIDS - always keep IDs regardless of ASGCT use (debug option)
-XX:-KeepJNIIDS - never keep IDs regardless of ASGCT use (prefer crash over leak)

Reply:
I like that approach, I think having the flexibility of toggling the jniID behaviour and having a debug option is beneficial.

Reply:
Do you want me to do this as part of this PR?

Reply:
yes, please do so